Changelog 175 (#157)
intellij-rust/intellij-rust.github.io: _posts/2022-07-25-changelog-175.markdown

---
layout: post
title: "IntelliJ Rust Changelog #175"
date: 2022-07-25 13:00:00 +0300
---

## New Features

> * [#8767] You can now paste JSON objects into Rust files and convert them to a set of Rust structs (by [@Kobzol])

Reviewer: Maybe it's worth to add all PRs related to this feature?

Reviewer: Also, it would be great to mention (and maybe show) that feature can add `serde` attributes

  {% include gif-img.html path="/assets/posts/changelog-175/json" w="700px" %}

* [#9012] Introduce new approach of project model reloading and detection changes in configuration files
  starting with 2022.2 major IDE release. Now the plugin detects changes in configuration files like
  `Cargo.toml` even if they are not saved into disk yet, takes into account
  [Cargo config](https://doc.rust-lang.org/cargo/reference/config.html),
  [toolchain](https://rust-lang.github.io/rustup/overrides.html#the-toolchain-file) and
  [build script](https://doc.rust-lang.org/cargo/reference/build-scripts.html) files and notifies you if
  the project model is not up-to-date via floating button in the top-right corner of the editor.
  You can select preferable way when the project model should be reloaded automatically in
  `Preferences | Build, Execution, Deployment | Build Tools`. At the same time, Cargo settings were
  moved to `Preferences | Build, Execution, Deployment | Build Tools | Cargo`

  {% include gif-img.html path="/assets/posts/changelog-175/new-cargo-model-auto-import" w="700px" %}

> * [#8868] Highlight macro-machinery-specific syntax in `macro_rules` definitions.

Reviewer: ```suggestion
* [#8868] Highlight macro-machinery-specific syntax in [delcrative macro](https://doc.rust-lang.org/book/ch19-06-macros.html#declarative-macros-with-macro_rules-for-general-metaprogramming) definitions.
```

Reviewer: nit: don't like `macro-machinery-specific syntax` collocation. Maybe rephrase it somehow?

Reviewer 2: Can't think of anything better `macro-machinery-specific` right now :(

  New highlighting should help to distinguish a syntax specific to macros from
  an expansion code (by [@daym])

  <img src="/assets/posts/changelog-175/macro-highlighting.png" alt="Macro highlighting" width="700px"/>

* [#8869] Enable by default some inspections for dependencies in `Cargo.toml`

    * `Invalid crate version` warns you about a non-existing crate version

      <img src="/assets/posts/changelog-175/invalid-crate-version.png" alt="Invalid crate version" width="700px"/>

    * `New crate version available` notifies you about a crate that could be updated.

      {% include gif-img.html path="/assets/posts/changelog-175/new-crate-version-available" w="700px" %}

      We tried to make the highlighting not too intrusive, because using an older version of a crate

>       may be quite normal. If you still don't like the inspection, you can simply [disable it][disable-inspection].

Reviewer: ```suggestion
      may be quite normal. If you still don't like the inspection, you can simply [disable it][disable-inspection]
```

* Improve detection of some errors:

    * [#9026] Improve detection of duplicate definitions.
      Add support for [`E0254`](https://doc.rust-lang.org/error-index.html#E0254),
      [`E0259`](https://doc.rust-lang.org/error-index.html#E0259) and
      [`E0260`](https://doc.rust-lang.org/error-index.html#E0260) errors

    * [#9020] Detect [E0430](https://doc.rust-lang.org/error-index.html#E0430)
      `The self import appears more than once in the list` compiler error

    * [#9000] Detect [E0742](https://doc.rust-lang.org/error-index.html#E0742)
      `Visibility is restricted to a module which isn’t an ancestor of the current item` compiler error

* [#9024] Support `View | Type Info` action for field shorthand syntax.

>   You can invoke `Type Info` action using <kbd>Ctrl</kbd> + <kbd>shift</kbd> + <kbd>P</kbd> shortcut

Reviewer: Maybe just collapse action path and its shortcut like `Support `View | Type Info` action (<kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>P</kbd>)` as we usually do?

> * [#7299] Fix debug pretty-printers to render `HashMap`/`HashSet` with MSVC LLDB on Windows

Reviewer: ```suggestion
* [#7299] Provide debugger pretty-printers to render `HashMap`/`HashSet` with MSVC LLDB on Windows
```

> * [#7299] Fix debug pretty-printers to render `HashMap`/`HashSet` with MSVC LLDB on Windows

Reviewer: I'd swap this item with "Type Info" and probably add a screenshot

## Performance Improvements

* [#9021], [#9016], [#9015], [#8968] Slightly speed up type inference

* [#8986] Speed up completion in procedural macros.
  Please, note that procedural macro expansion is experimental. If you want to try it,

>   enable `org.rust.cargo.evaluate.build.scripts` and `org.rust.macros.proc` experimental features

Reviewer: ```suggestion
  enable `org.rust.cargo.evaluate.build.scripts` and `org.rust.macros.proc` [experimental features](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-faq.html#experimental-features)
```

## Fixes

* [#8986] Try to fix up rust syntax before calling an attribute procedural macro placed on a function.
  This fixes highlighting blinking during typing in a body of a function with a proc attribute macro on it

* [#9090] Limit the maximal `#![recursion_limit]` attribute value to 1000 when expanding macros.
  This fixes possible stack overflows

> * [#9085] Fix possible deadlock during indexing

Reviewer: nit: I'd move it into performance section

* [#9058] Fixes procedural macro expansion on Rust beta

* [#9009] Fix ["Inline function"] refactoring dialog when function has single usage in another module

> * [#8951] Do not offer to remove unused variables in places where they are required (by [@Kobzol])

Reviewer: Actually, it's quite unclear phrase, from my point of view. Let's rephrase it somehow

## Internal Improvements

* [#9070] Migrate usages of [Kotlin UI DSL version 1][dsl1] to [the second version][dsl2]

* [#9025] Make 2022.2 platform default for development

* [#9067] Fixes exception on opening `Configure features` popup in Cargo.toml (by [@Stzx])

>   The popup can be observed only if `org.rust.cargo.features.settings.gutter` experimental feature is enabled

Reviewer: Maybe move it to the end of "Fixes" section?

>   The popup can be observed only if `org.rust.cargo.features.settings.gutter` experimental feature is enabled

Reviewer: ```suggestion
  The popup can be observed only if `org.rust.cargo.features.settings.gutter` [experimental feature](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-faq.html#experimental-features) is enabled
```
